### PR TITLE
Feature/cpp protocol gateway

### PR DESCRIPTION
**Pull request by halvfigur**
Implement ProtocolGateway

### CODE REVIEW

**Comment by halvfigur** [comment]
This branch is based off of feature/get_features_and_bugfixes which is also in review.

**Comment by JochenKienzle** [comment]
I have added @llakie to review the implementation on API / interface level.
@llakie Is this ok for you? I would look at the c++ implementation (but less on the the api / interface level)

File: src/sdk/cpp/lib/src/util.cpp [hunk 1]
```diff
@@ -120,5 +120,36 @@ std::string GenerateUUID() {
...
+
+TopicExprMatcher::TopicExprMatcher(std::string topic_expr) {
+    auto size = topic_expr.size();
+
+    // '#' may only appear at the end of the expression
+    if (topic_expr[size - 1] == '#') {
+        topic_expr.replace(size - 1, 1, ".*");
+    }
+
+    // '$' must be escaped
```
Comment (halvfigur) [inline src/sdk/cpp/lib/src/util.cpp]: Period "." must also be escaped.

File: src/sdk/cpp/adapters/mqtt/src/mqtt_adapter.cpp [hunk 2]
```diff
@@ -137,41 +158,42 @@ void MqttClient::Run() {
...
-        next_state_ = state;
-    }
-}
+void MqttProtocolAdapter::Publish(const std::string& topic, const std::string& data, const PublishOptions&) {
+    auto full_topic = topic_ns_ + topic;
 
-void MqttClient::Publish(const std::string& topic, const std::string& data) {
     log::Debug() << "Publishing message.";
-    log::Debug() << "\ttopic: '" << topic << "'";
+    log::Debug() << "\ttopic: '" << full_topic << "'";
```
Comment (JochenKienzle) [inline src/sdk/cpp/adapters/mqtt/src/mqtt_adapter.cpp]: I think the logs (all) have to be merged with the previous PR (https://github.com/GENIVI/iot-event-analytics/pull/79)

File: src/sdk/cpp/examples/basic/basic.cpp [hunk 3]
```diff
@@ -31,14 +36,12 @@ using iotea::core::schema::rule_ptr;
...
 
-constexpr char SERVER_ADDRESS[] = "tcp://localhost:1883";
-
 class MyService : public Talent {
    public:
     MyService() : Talent("cpp-basic-talent") { }
 
     rule_ptr OnGetRules() const override {
-        return Change("temp", "kuehlschrank");
+        return Change("anyfeature", "anytype");
```
Comment (JochenKienzle) [inline src/sdk/cpp/examples/basic/basic.cpp:43]: bye bye to 'Kuehlschrank' ;-)

File: src/sdk/cpp/lib/CMakeLists.txt [hunk 4]
```diff
@@ -34,6 +34,8 @@ endif()
 set (THREADS_PREFER_PTHREAD_FLAG ON)
 find_package(Threads REQUIRED)
 
+find_library(LIBDL dl)
```
Comment (JochenKienzle) [inline src/sdk/cpp/lib/CMakeLists.txt:37]: Does this imply that we are 'limited' to dynamic linking?

File: src/sdk/cpp/lib/include/call.hpp [hunk 5]
```diff
@@ -315,20 +318,15 @@ class PreparedFunctionReply {
...
      * @param talent_id The ID of the talent to send the reply to
      * @param feature The name of the feature
-     * @param subject The subject of the original event resulting in the
-     * function call
-     * @param channel_id The ID of the calling talent's channel_id
-     * @param call_id The unique call IDs
+     * @param event The event that triggered the call
      * @param return_topic The topic on which the reply should be posted
-     * @param publisher A pointer to a Publisher
+     * @param gatewy A pointer to a GatewayProtocl
```
Comment (JochenKienzle) [inline src/sdk/cpp/lib/include/call.hpp:323]: typo gatew(a)y

File: src/sdk/cpp/lib/include/event.hpp [hunk 6]
```diff
@@ -264,14 +251,16 @@ class Event {
...
+     * @param features The feature object contained within the event.
      * @param type The name of the type associated with the event.
      * @param instance The name of the instance associated with the event.
      * @param return_topic The name of the topic to send replies on (if any).
      * @param when The point in time when the event was emitted.
      */
     Event(const std::string& subject, const std::string& feature, const json& value,
-          const std::string& type = "default", const std::string& instance = "default",
-          const std::string& return_topic = "", int64_t when = GetEpochTimeMs());
+          const json& features, const std::string& type = "default",
```
Comment (JochenKienzle) [inline src/sdk/cpp/lib/include/event.hpp:273]: idea/improvement: "default" could be an enum / const

File: src/sdk/cpp/lib/src/protocol_gateway.cpp [hunk 7]
```diff
@@ -0,0 +1,215 @@
...
+}
+
+/////////////////////
+// ProtocolGateway //
+/////////////////////
+const PublishOptions ProtocolGateway::DefaultPublishOptions{false, ""};
+
+const SubscribeOptions ProtocolGateway::DefaultSubscribeOptions{false, ""};
+
+void ProtocolGateway::ValidateConfig(const json& config, bool platform_proto_only) {
```
Comment (JochenKienzle) [inline src/sdk/cpp/lib/src/protocol_gateway.cpp:100]: Does it make sense to add a flag to activate / deactivate the validation (e.g. activated for debug, deactivated for release mode). Can be done in an upcoming PR :)

**Review by JochenKienzle — CHANGES_REQUESTED** [review]
Overall looks very good 👍  I like the state-machine based approach in the gateway adapter 👍 

Some small findings (typo, merges) and one question regarding static/dynamic linking

File: src/sdk/cpp/lib/CMakeLists.txt (hunk 4 above)
Comment (halvfigur) [reply src/sdk/cpp/lib/CMakeLists.txt:37]: Yes

File: src/sdk/cpp/lib/include/call.hpp (hunk 5 above)
Comment (halvfigur) [reply src/sdk/cpp/lib/include/call.hpp:323]: I'm going to update the documentation and provide unit tests in a different commit. If it's ok with you I'd like to merge this and correct it in the next PR.

File: src/sdk/cpp/lib/src/protocol_gateway.cpp (hunk 7 above)
Comment (halvfigur) [reply src/sdk/cpp/lib/src/protocol_gateway.cpp:100]: Maybe, this is cheap check and modelled after what the Python SDK is doing.

**Review by JochenKienzle — APPROVED** [review]

**Pull request merged**